cpp/include/raft/spatial/knn/ivf_pq_types.hpp

```diff
@@ -0,0 +1,428 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "common.hpp"
+
+#include <raft/core/device_mdarray.hpp>
+#include <raft/core/error.hpp>
+#include <raft/core/host_mdarray.hpp>
+#include <raft/distance/distance_types.hpp>
+#include <raft/util/integer_utils.hpp>
+
+#include <rmm/mr/device/managed_memory_resource.hpp>
+
+#include <type_traits>
+
+namespace raft::spatial::knn::ivf_pq {
+
+/** A type for specifying how PQ codebooks are created. */
+enum class codebook_gen {
+  PER_SUBSPACE = 0,
+  PER_CLUSTER  = 1,
+};
+
+struct index_params : knn::index_params {
```

Comment:
FYI, `randomRotation` bool param was there before I ported the implementation to GPU. I removed it mainly to simplify my work, assuming that it was needed only because the CPU version could be very slow for large dimensions.

Reply:
Update: added `force_random_rotation` parameter to change the behavior only in case of the square transform.